src/borg/archive.py

```diff
@@ -747,29 +747,60 @@ def compare_and_extract_chunks(self, item, fs_path, *, st, pi=None):
             # Second pass: Update file
             with backup_io("open"):
                 fs_file = open(fs_path, "rb+")
-            with fs_file:
                 for fs_chunk, item_chunk in zip(fs_chunks, item.chunks):
                     if fs_chunk == item_chunk:
                         with backup_io("seek"):
-                            fs_file.seek(item_chunk.size, 1)
+                            fs_file.seek(item_chunk.size, os.SEEK_CUR)
+                            size = item_chunk.size
                     else:
                         chunk_data = next(chunk_data_iter)
 
                         with backup_io("write"):
                             fs_file.write(chunk_data)
+                            size = len(chunk_data)
                     if pi:
-                        pi.show(increase=len(chunk_data), info=[remove_surrogates(item.path)])
+                        pi.show(increase=size, info=[remove_surrogates(item.path)])
 
-                final_size = fs_file.tell()
                 with backup_io("truncate_and_attrs"):
-                    fs_file.truncate(item.size)
+                    pos = item_chunks_size = fs_file.tell()
+                    fs_file.truncate(pos)
                     fs_file.flush()
+
+                    if not self.noacls:
+                        try:
+                            # Clear ACLs by setting empty ACL entries
+                            acl_set(fs_path, {"acl_entries": []}, self.numeric_ids, fd=fs_file.fileno())
+                        except OSError as e:
+                            if e.errno != errno.ENOTSUP:  # Ignore if ACLs not supported
+                                raise
+                    if not self.noflags:
+                        try:
+                            # Clear all BSD flags
+                            set_flags(fs_path, 0, fd=fs_file.fileno())
+                        except OSError:
+                            pass
+                    if not self.noxattrs:
+                        try:
+                            # Clear ALL xattrs
+                            attrs = xattr.listxattr(fs_file.fileno(), follow_symlinks=False)
+                            for attr in attrs:
+                                try:
+                                    xattr.setxattr(fs_file.fileno(), attr, b"", follow_symlinks=False)
+                                except OSError as e:
+                                    if e.errno != xattr.ENOATTR:  # Ignore if already removed
+                                        raise
+                        except OSError as e:
+                            if e.errno != errno.ENOTSUP:  # Ignore if xattrs not supported
+                                raise
                     self.restore_attrs(fs_path, item, fd=fs_file.fileno())
```

Comment:
you could create a method clear_attrs close to the definition of restore_attrs and move that code you added above to there.